[Security] Throw a forbidden exception when the IsCsrfTokenValid attribute fails

eltharin opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #57343
| License       | MIT

`IsCsrfTokenValidAttributeListener` threw `Security\Core\Exception\InvalidCsrfTokenException`, which extends `AuthenticationException`. `Security\Http\Firewall\ExceptionListener` matches that before anything else and routes it to `startAuthentication()`, so a CSRF failure inside a controller redirected to the login page, or returned 401 when the firewall has no entry point.

The Core exception is the right one for the authentication-time check in `CsrfProtectionListener`, where the login form's token is verified. It is the wrong one for a controller attribute, which is reached by a caller who is already identified. That is the distinction @stof drew when the attribute was added.

A new `Security\Http\Exception\InvalidCsrfTokenException` extending `HttpException` with a 403 status replaces it, matching what `IsGrantedAttributeListener` already throws when a status code is configured.

Measured end to end with a real kernel, the real `ControllerAttributesListener`, the real firewall `ExceptionListener` and `debug=false`:

| firewall | before | after |
| --- | --- | --- |
| `form_login` entry point | 302 to /login | 403 |
| no entry point | 401 | 403 |
| `http_basic` entry point | 401 with `WWW-Authenticate` | 403 |
| no firewall exception listener | 401, from `#[WithHttpStatus]` | 403 |

No path produces a 500, and none produces a 401 afterwards. The exception message is unchanged and is not exposed any differently: hidden by both renderers when `debug=false`, and the JSON problem detail moves from `"Unauthorized"` to `"Forbidden"`, which is the status text rather than the message.

Changed during review:

* the CHANGELOG entry moved from `SecurityBundle` to `Security\Http`, where the changed class lives, and into the 8.2 section. It had landed in the 7.3 section after rebasing. This is what @kbond asked for;
* added the `[BC BREAK]` entry in `UPGRADE-8.2.md`. Two things need it: the old and new exceptions share no ancestor below `\RuntimeException`, so an application catching the `Security\Core` one for this case silently stops matching, and stateless firewalls move from 401 to 403;
* added a test for the behaviour rather than only the class name. The suite asserted which exception was thrown and nothing else, so reparenting the class onto something the firewall matches would have kept it green. The new test asserts the 403 status and that the exception is not an `AuthenticationException`, and it fails on a clean 8.2;
* the redundant explicit message at the call site is gone, and the two same-named exceptions no longer carry identical docblocks.

Left as is deliberately: the parent stays `HttpException` rather than the `BadRequestHttpException` suggested in review, because that class means 400 and the constructor would have to bypass its parent to force 403. @webda2l's request for a configurable `statusCode` like `IsGranted` is not implemented here.

eltharin commented:
SomeOne can help me to say what in 8.2 high-deps the test is calling old files and wait for a redirect ?

eltharin committed:
change Parent Exception

clear old getMessage function

eltharin commented:
seems test fail because it take old test with depedencies ??

stof commented:
The `InvalidCsrfTokenException` in `security/core` is meant to be used for invalid CSRF tokens during the authentication. In that case, extending AuthenticationException is the thing we want.

To me, the issue is that the `IsCsrfTokenValid` attribute should not be reusing that exception.

stof commented:
Side note, the `IsCsrfTokenValid` attribute should probably not have been added in `security/http`. It should probably have been part of `security/csrf` instead.

eltharin commented:
So I can keep in place InvalidCsrfTokenException and create another on in Security/Http extending BadRequestException and link it to the Exception throw in IsCsrfTokenValidAttributeListener ? 
thanks for review

eltharin committed:
follow review

eltharin committed:
revert test

eltharin commented:
change is OK, do I move IsCsrfTokenValidAttributeListener  in Security/Csrf ?

eltharin committed:
changelog

eltharin committed:
fabbot changes

kbond reviewed (commented):
I think we should move the attribute/listener

kbond commented on src/Symfony/Bundle/SecurityBundle/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ CHANGELOG
  * Add ability to fetch LDAP roles
  * Add `OAuth2TokenHandlerFactory` for `AccessTokenFactory`
  * Add discovery support to `OidcTokenHandler` and `OidcUserInfoTokenHandler`
+ * Change Exception thrown by IsCsrfTokenValid Attribute
```
This should be in `security-csrf` & `security-http`

kbond commented on src/Symfony/Component/Security/Csrf/EventListener/IsCsrfTokenValidAttributeListener.php:
```diff
@@ -9,17 +9,17 @@
  */
 
-namespace Symfony\Component\Security\Http\EventListener;
+namespace Symfony\Component\Security\Csrf\EventListener;
```
Unfortunately, it isn't this simple, we need to keep it in both places, deprecating the one in http. See #46094 for how we did this with the `Security` helper.

kbond reviewed (commented):
Thanks for working on this @eltharin!

Before going further with *moving* the listener/attribute to security-csrf, let's see what other @symfony/mergers think. Maybe we should just focus this PR on fixing the bug and the move in a follow-up PR (or not at all).

eltharin commented:
Ok I revert the move and wait and see

smnandre commented:
I **strongly** disagree here.

Regardless of whether the user is logged in, CSRF is **100% an authentication** problem.

> CSRF is an attack that tricks the victim into submitting a malicious request. It inherits the identity and privileges of the victim…
([OWASP](https://owasp.org/www-community/attacks/csrf))

The CSRF token system  -and by extension the `IsCsrfTokenValid` attribute- have one purpose: to ensure a request was intentionally sent by the user.

If the listener can’t validate the token, it can’t confirm that the user who sent the request is who they claim to be (again: regardless of its login status). 

And not being able to confirm the identity or intent behind a request is, by definition, an authentication failure.

eltharin commented:
> I **strongly** disagree here.
> 
> Regardless of whether the user is logged in, CSRF is **100% an authentication** problem.
> 
> > CSRF is an attack that tricks the victim into submitting a malicious request. It inherits the identity and privileges of the victim…
> > ([OWASP](https://owasp.org/www-community/attacks/csrf))
> 
> The CSRF token system -and by extension the `IsCsrfTokenValid` attribute- have one purpose: to ensure a request was intentionally sent by the user.
> 
> If the listener can’t validate the token, it can’t confirm that the user who sent the request is who they claim to be (again: regardless of its login status).
> 
> And not being able to confirm the identity or intent behind a request is, by definition, an authentication failure.

We can keep 403 code error, the really problem is the login redirection.

nicolas-grekas commented on src/Symfony/Component/Security/Csrf/Exception/InvalidCsrfTokenException.php:
```diff
@@ -0,0 +1,23 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Csrf\Exception;
```
putting this exception in the CSRF component makes this more complex than needed
given we went with putting the attribute in the security-http component (which kinda follows the idea that this attribute is for integration with the HttpKernel component / controllers), I'd put the exception in the security-http.
In order to still default to a 403, this could work, isn't it?
would this still trigger the redirection?

```php
namespace Symfony\Component\Security\Http\Exception;

use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
use Symfony\Component\HttpKernel\Exception\HttpException;

class InvalidCsrfTokenException extends BadRequestHttpException
{
    public function __construct(string $message = '', ?\Throwable $previous = null, int $code = 0, array $headers = [])
    {
        HttpException::__construct(403, $message, $previous, $headers, $code);
    }
}
```

eltharin replied:
yes it's possible to put it in security-http, it need to extends another class than AuthenticationException as actual (bad) behaviour.

eltharin committed:
change csrf => http

eltharin committed:
[Security] Throw a forbidden exception when the IsCsrfTokenValid attribute fails

The listener threw Security\Core\InvalidCsrfTokenException, which extends
AuthenticationException. The firewall ExceptionListener matches that first and
routes it to startAuthentication(), so a CSRF failure inside a controller
redirected to the login page, or returned 401 when the firewall has no entry
point. The Core exception is right for the authentication-time check in
CsrfProtectionListener; it is not right for a controller attribute reached by
an already identified caller.

A new Security\Http exception extending HttpException with a 403 status
replaces it, which is also what IsGrantedAttributeListener does when a status
code is configured.

kbond reviewed (approved):
This looks great, and the status code can now be properly overridden in `framework.exceptions`!
